go/vt/vtgate/engine/scalar_aggregation.go

```diff
@@ -0,0 +1,193 @@
+	qr, err := sa.Input.GetFields(vcursor, bindVars)
+	if err != nil {
+		return nil, err
+	}
+	qr = &sqltypes.Result{Fields: convertFields(qr.Fields, sa.PreProcess, sa.Aggregates)}
+	return qr.Truncate(sa.TruncateColumnCount), nil
+}
+
+// NeedsTransaction implements the Primitive interface
+func (sa *ScalarAggregate) NeedsTransaction() bool {
+	return sa.Input.NeedsTransaction()
+}
+
+// TryExecute implements the Primitive interface
+func (sa *ScalarAggregate) TryExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	result, err := vcursor.ExecutePrimitive(sa.Input, bindVars, wantfields)
+	if err != nil {
+		return nil, err
+	}
+	out := &sqltypes.Result{
+		Fields: convertFields(result.Fields, sa.PreProcess, sa.Aggregates),
+	}
+
+	var resultRow []sqltypes.Value
+	var curDistincts []sqltypes.Value
+	for _, row := range result.Rows {
+		if resultRow == nil {
+			resultRow, curDistincts = convertRow(row, sa.PreProcess, sa.Aggregates)
+			continue
+		}
+		resultRow, curDistincts, err = merge(result.Fields, resultRow, row, curDistincts, sa.Collations, sa.Aggregates)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if resultRow == nil {
+		// When doing aggregation without grouping keys, we need to produce a single row containing zero-value for the
+		// different aggregation functions
+		resultRow, err = sa.createEmptyRow()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		resultRow, err = convertFinal(resultRow, sa.Aggregates)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	out.Rows = [][]sqltypes.Value{resultRow}
+	return out, nil
+}
+
+// TryStreamExecute implements the Primitive interface
+func (sa *ScalarAggregate) TryStreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	cb := func(qr *sqltypes.Result) error {
+		return callback(qr.Truncate(sa.TruncateColumnCount))
+	}
+	var current []sqltypes.Value
+	var curDistincts []sqltypes.Value
+	var fields []*querypb.Field
+	fieldsSent := false
+
+	err := vcursor.StreamExecutePrimitive(sa.Input, bindVars, wantfields, func(result *sqltypes.Result) error {
+		if len(result.Fields) != 0 && !fieldsSent {
+			fields = convertFields(result.Fields, sa.PreProcess, sa.Aggregates)
+			if err := cb(&sqltypes.Result{Fields: fields}); err != nil {
+				return err
+			}
+			fieldsSent = true
+		}
+
+		// this code is very similar to the TryExecute method
+		for _, row := range result.Rows {
+			if current == nil {
+				current, curDistincts = convertRow(row, sa.PreProcess, sa.Aggregates)
+				continue
+			}
+			var err error
+			current, curDistincts, err = merge(fields, current, row, curDistincts, sa.Collations, sa.Aggregates)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return cb(&sqltypes.Result{Rows: [][]sqltypes.Value{current}})
```

Comment:
the original oa code has a check for 
`current != nil `
does this make a difference if you are not performing the nil check on current?